src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -228,24 +233,41 @@ final class ArrayBufferView[A](val array: Array[AnyRef], val length: Int) extend
 /** An object used internally by collections backed by an extensible Array[AnyRef] */
 object RefArrayUtils {
 
+  // We chose 8 as a good size for the minimum array
+  private[this] val minimalLength: Long = 8L
+
   def ensureSize(array: Array[AnyRef], end: Int, n: Int): Array[AnyRef] = {
+    if (end < 0 || n < 0)
+      throw new Exception(s"Collections can not have a negative size")
+    if (end == Int.MaxValue)
+      throw new Exception(s"Collections can not have more than ${Int.MaxValue} elements")
     // Use a Long to prevent overflows
     val arrayLength: Long = array.length
     def growArray = {
-      var newSize: Long = math.max(arrayLength * 2, 8)
+      var newSize: Long = math.max(arrayLength * 2, minimalLength)
       while (n > newSize)
         newSize = newSize * 2
       // Clamp newSize to Int.MaxValue
-      if (newSize > Int.MaxValue) {
-        if (end == Int.MaxValue) throw new Exception(s"Collections can not have more than ${Int.MaxValue} elements")
+      if (newSize > Int.MaxValue)
         newSize = Int.MaxValue
-      }
 
       val newArray: Array[AnyRef] = new Array(newSize.toInt)
       Array.copy(array, 0, newArray, 0, end)
       newArray
     }
-    if (n <= arrayLength) array else growArray
+    def shrinkArray = {
+      val targetSize = math.max(2*n, minimalLength)
+      var newSize: Long = arrayLength
+      while (newSize / 2 > targetSize)
+        newSize = newSize / 2
+      val newArray: Array[AnyRef] = new Array(newSize.toInt)
+      Array.copy(array, 0, newArray, 0, end)
+      newArray
+    }
+
+    if (n > arrayLength) growArray
+    else if (n <= arrayLength / 4) shrinkArray
```

Comment:
Putting the logic for shrinking into `ensureSize` doesn't look like a good idea to me. This should be a separate operation. Otherwise an explicit call to `ensureSize` with a large size (before adding lots of elements) will grow the array and the next call to `addOne` will shrink it back down.